src/OrchardCore/OrchardCore.DisplayManagement/TagHelpers/BaseShapeTagHelper.cs

```diff
@@ -159,10 +164,36 @@ public override async Task ProcessAsync(TagHelperContext tagHelperContext, TagHe
 
             await output.GetChildContentAsync();
 
+            await OnDisplayingAsync(shape, tagHelperContext, output);
+
             output.Content.SetHtmlContent(await _displayHelper.ShapeExecuteAsync(shape));
 
             // We don't want any encapsulating tag around the shape
             output.TagName = null;
         }
+
+        /// <summary>
+        /// Extension point that may be invoked in a derived class to alter the <paramref name="properties"/> before the
+        /// shape is created with them.
+        /// </summary>
+        protected virtual Task OnPropertiesAddedAsync(
+            IDictionary<string, object> properties,
+            TagHelperContext tagHelperContext,
+            TagHelperOutput output) =>
+            Task.CompletedTask;
+
+        /// <summary>
+        /// Extension point that may be invoked in a derived class to alter the <paramref name="shape"/> before it's
+        /// executed and rendered into HTML.
+        /// </summary>
+        /// <param name="shape"></param>
+        /// <param name="tagHelperContext"></param>
+        /// <param name="output"></param>
+        /// <returns></returns>
+        protected virtual Task OnDisplayingAsync(
```

Comment:
IMO You don't need this.

Reply:
It would've prevented future similar PRs, by allowing devs to extend the shape's functionality in their own derived classes without having to modify this base class. But ok, for the sake of finishing this PR I've removed it.

Reply:
`Shape` itself versatile and extensible, and  the objective  can be achieved  by chaining `Shape` object.  Also method `OnDisplayingAsync` is misleading here, shape itself has `Displaying` events and actually content are not being displayed here.

Reply:
ok